### ISSUE TITLE
why is XGBoost giving me seriously biased predictions with small nrounds?

### ISSUE THREAD

**Issue by metariat** [dmlc/xgboost#3080]
Package used (python/R/jvm/C++): R

`xgboost` version used: 0.6-4

If you are using R package, please provide

1. The R `sessionInfo()`
2. The command to install `xgboost` if you are not installing from source: install.packages('xgboost')

## Steps to reproduce

I can not put my company data online, but I can provide a reproducible example here.

We're modelling Insurance's frequency using Poisson distribution with exposure as offset.

Here in this example, we want to model the number of claim `Claims` ($y_i$) with exposure `Holders` ($e_i$)

In the traditional GLM model, we can dirrectly model $y_i$ and put $e_i$ in the offset term. This option is not available in `xgboost`. So the alternative is to model the rate $\frac{y_i}{e_i}$, and put $e_i$ as a the weight term ([reference][1])

When I do that with a lot of iteractions, the results are coherent ($\sum y_i = \sum \hat{y_i}$). But it is not the case when `nrounds = 5`. I think that the equation $\sum y_i = \sum \hat{y_i}$ must be satisfied after the very first iteration. 

The following code is an extreme example for the sake of reproducibility. In my real case, I performed a CV on the training set (optimizing MAE), I obtained `nrounds = 1200`, training MAE = testing MAE. Then I re-run a xgboost on the whole data set with 1200 iteration, I see that $\sum y_i \ne \sum \hat{y_i}$ by a large distance, this doesn't make sense, or am I missing something?

So my questions are:

 1. Am I correctly specify parameters for Poisson regression with offset in `xgboost`?
 2. Why such a high bias at the first iterations?
 3. Why after tuning `nrounds` using xgb.cv, I still have high bias?

```
    library(MASS)
    library(caret)
    library(xgboost)
    library(dplyr)
    
    #-------- load data --------#
    data(Insurance)
    
    #-------- data preparation --------#
    
    #small adjustments
    Insurance$rate = with(Insurance, Claims/Holders)
    temp<-dplyr::select(Insurance,District, Group, Age, rate)
    temp2= dummyVars(rate ~ ., data = temp, fullRank = TRUE) %>% predict(temp)
    
    #create xgb matrix
    xgbMatrix <- xgb.DMatrix(as.matrix(temp2), 
                             label = Insurance$Claims)
    setinfo(xgbMatrix, "base_margin",log(Insurance$Holders))
    
    
    #-------------------------------------------#
    #      First model with small nround
    #-------------------------------------------#
    bst.1 = xgboost(data    = xgbMatrix, 
                    objective ='count:poisson', 
                    nrounds   = 5)
    
    
    pred.1 = predict(bst.1, xgbMatrix)
    sum(Insurance$Claims) #3151
    sum(pred.1) #12650.8 fails
    
    
    
    #-------------------------------------------#
    #      Second model with more iteractions
    #-------------------------------------------#
    
    bst.2 = xgboost(data    = xgbMatrix, 
                    objective = 'count:poisson', 
                    nrounds   = 100)
    
    
    pred.2 = predict(bst.2, xgbMatrix)
    sum(Insurance$Claims) #3151
    sum(pred.2) #same
```

**Comment by Laurae2** [comment]
@metariat You need to change the learning rate (`eta`) for gradient descent in xgboost.

Same as with GLMs (they are based on gradient descent optimization or similar), if you do not provide enough iterations for xgboost, the results will be far off the expected value.

**Comment by khotilov** [comment]
To add to the above, Poisson objective also contains an additional trick to make its Hessian numerically more stable by trading off some optimality:
https://github.com/dmlc/xgboost/blob/077abb35cd04f18b16faf1e1744c7bbcbbdd71e6/src/objective/regression_obj.cc#L164
The `max_delta_step` (default values is empirically set to 0.7 for this objective) that is artificially added to the exponent in hessian, it "majorizes" it within a +-max_delta_step step (as in Majorization Minimization). Thus, for this objective you have to train for more iterations than it is typically needed for the others.

**Comment by metariat** [comment]
In fact, the example above may be not representative of my problem. I fitted a Poisson model with offset to my data. I choosed nrounds by doing xgb.cv, I plotted the two curve training / validation error, they are identical. Then I used the nrounds found previously to refit the model on the whole training data. I have two strange behaviours: sum of observations is not equal to sum of predictions, and the xgboost largely overfits my training data (while the training and cv errors are identical). I have learning_rate = 0.01 and nrounds = 1200.
